### ISSUE TITLE
Cache Configuration for Limiting Entries Fails on Startup

### ISSUE THREAD

**Issue by jfrosch** [grails-plugins/grails-cache#62]
After getting the error in the client app, I created a new project in isolation to reproduce it. Steps to recreate:

- Created new rest-api project
`curl -O start.grails.org/VerifyCacheConfig.zip -d version=3.3.6 -d profile=rest-api`
- Added Cache plugin dependency which loaded v4.0.0 of plugin
`compile "org.grails.plugins:cache"`
- Added this configuration:
```---
grails:
    cache:
        cacheManager: GrailsConcurrentLinkedMapCacheManager
        caches:
            foo:
                maxCapacity: 5000
            bar:
                maxCapacity: 10000
```
- Got this stack trace on `bootrun` startup:
```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'grailsCacheConfiguration': Could not bind properties to CachePluginConfiguration (prefix=grails.cache, ignoreInvalidFields=false, ignoreUnknownFields=true, ignoreNestedProperties=false); nested exception is org.springframework.beans.NullValueInNestedPathException: Invalid property 'caches[foo]' of bean class [grails.plugin.cache.CachePluginConfiguration]: Could not instantiate property type [grails.plugin.cache.CachePluginConfiguration$CacheConfig] to auto-grow nested property path; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [grails.plugin.cache.CachePluginConfiguration$CacheConfig]: Is it an abstract class?; nested exception is java.lang.InstantiationException: grails.plugin.cache.CachePluginConfiguration$CacheConfig
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:336)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:292)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:409)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1626)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:308)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:761)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:867)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:543)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:84)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:393)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:380)
	at grails.boot.GrailsApp$run.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:136)
	at verifycacheconfig.Application.main(Application.groovy:8)
Caused by: org.springframework.beans.NullValueInNestedPathException: Invalid property 'caches[foo]' of bean class [grails.plugin.cache.CachePluginConfiguration]: Could not instantiate property type [grails.plugin.cache.CachePluginConfiguration$CacheConfig] to auto-grow nested property path; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [grails.plugin.cache.CachePluginConfiguration$CacheConfig]: Is it an abstract class?; nested exception is java.lang.InstantiationException: grails.plugin.cache.CachePluginConfiguration$CacheConfig
	at org.springframework.beans.AbstractNestablePropertyAccessor.newValue(AbstractNestablePropertyAccessor.java:931)
	at org.springframework.beans.AbstractNestablePropertyAccessor.createDefaultPropertyValue(AbstractNestablePropertyAccessor.java:899)
	at org.springframework.beans.AbstractNestablePropertyAccessor.setDefaultValue(AbstractNestablePropertyAccessor.java:887)
	at org.springframework.beans.AbstractNestablePropertyAccessor.getNestedPropertyAccessor(AbstractNestablePropertyAccessor.java:851)
	at org.springframework.beans.AbstractNestablePropertyAccessor.getPropertyAccessorForPropertyPath(AbstractNestablePropertyAccessor.java:825)
	at org.springframework.beans.AbstractNestablePropertyAccessor.setPropertyValue(AbstractNestablePropertyAccessor.java:268)
	at org.springframework.boot.bind.RelaxedDataBinder$RelaxedBeanWrapper.setPropertyValue(RelaxedDataBinder.java:726)
	at org.springframework.beans.AbstractPropertyAccessor.setPropertyValues(AbstractPropertyAccessor.java:95)
	at org.springframework.validation.DataBinder.applyPropertyValues(DataBinder.java:860)
	at org.springframework.validation.DataBinder.doBind(DataBinder.java:756)
	at org.springframework.boot.bind.RelaxedDataBinder.doBind(RelaxedDataBinder.java:137)
	at org.springframework.validation.DataBinder.bind(DataBinder.java:741)
	at org.springframework.boot.bind.PropertiesConfigurationFactory.doBindPropertiesToTarget(PropertiesConfigurationFactory.java:287)
	at org.springframework.boot.bind.PropertiesConfigurationFactory.bindPropertiesToTarget(PropertiesConfigurationFactory.java:250)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:331)
	... 24 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [grails.plugin.cache.CachePluginConfiguration$CacheConfig]: Is it an abstract class?; nested exception is java.lang.InstantiationException: grails.plugin.cache.CachePluginConfiguration$CacheConfig
	at org.springframework.beans.BeanUtils.instantiate(BeanUtils.java:80)
	at org.springframework.beans.AbstractNestablePropertyAccessor.newValue(AbstractNestablePropertyAccessor.java:926)
	... 38 common frames omitted
Caused by: java.lang.InstantiationException: grails.plugin.cache.CachePluginConfiguration$CacheConfig
	at org.springsource.loaded.ri.Exceptions.instantiation(Exceptions.java:118)
	at org.springsource.loaded.ri.ReflectiveInterceptor.jlClassNewInstance(ReflectiveInterceptor.java:1059)
	at org.springframework.beans.BeanUtils.instantiate(BeanUtils.java:77)
	... 39 common frames omitted
```

**Comment by jfrosch** [comment]
Fwiw, I just tried the same using 5.0.0.BUILD-SNAPSHOT and got the same error.

**Comment by jfrosch** [comment]
I tried making the inner class in CachePluginConfiguration static and found this fixed the problem. 

I then discovered this appears to be a fix applied to 4.0.3.BUILD-SNAPSHOT.  I recommend 4.0.3 become a released version as soon as practical.

**Comment by jfrosch** [comment]
I thought the fix was in 4.0.3.BUILD-SNAPSHOT, but had inadvertently used the patched change I was testing which I had published into my local Maven repo.

This issue is still present in 4.0.3-BUILD-SNAPSHOT and 5.0.0-BUILD-SNAPSHOT.

**Comment by jameskleeh** [comment]
Fixed in 4.0.3
